[mod] rework container deployment

inetol opened the pull request:
This PR supersedes and is part of https://github.com/searxng/searxng/pull/4707

Closes #484

inetol committed:
[mod] rework container deployment

container.yml will run after integration.yml COMPLETES successfully and in master branch.

Style changes, cleanup and improved integration with CI by leveraging the use of
shared cache between all workflows.

* Podman is now supported to build the container images (Docker also received a refactor, merging both build and buildx)
* Container images are being built by Buildah instead of Docker BuildKit.
* Container images are tested before release.
* Splitting "modern" (amd64 & arm64) and "legacy" (armv7) arches on different Dockerfiles allowing future optimizations.

inetol commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+        pid_logs=$!
+
+        # Wait until container is ready
+        sleep 5
+
+        curl -vf --max-time 5 "http://localhost:8080/healthz"
```
No

return42 reviewed (commented):
> @inetol requested your review on: https://github.com/searxng/searxng/pull/4764 [mod] rework container deployment.

I'm sorry, I can't review since I don't have any docker knowledge.

dalf commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+    esac
+    build_msg CONTAINER "Selected platform: $platform"
+
+    # Check if podman is installed
```
Nitpicking: I would move all requirements at one top of the function, just after

```sh
    if [ "$GITHUB_ACTIONS" != "true" ]; then
        die 1 "This command is intended to be run in GitHub Actions"
    fi
```

(same for `container.push`)

inetol replied:
392a033913ca9496935ebe66bac061badda79d01

dalf commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+        platform="linux/$arch"
+        ;;
+    "ARMV7" | "armhf" | "armv7l" | "armv7")
+        dockerfile="armv7.dockerfile"
```
Nitpicking: why not `Dockerfile.armv7`, so this is next to `Dockerfile` when the file list is sorted?

inetol replied:
I put it as `.dockerfile` because the IDE highlights it correctly with that extension, which doesn't happen with `dockerfile.`

I will try to move the Dockerfiles to dedicated folders instead of having it in the project root.

dalf reviewed (approved):
LGTM

I haven't test `container.push`, but it makes sense.
I'm not sure why this is draft?

Thank you.

inetol committed:
move Dockerfiles to container folder

`./container/` folder is the new folder for container related files including new Dockerfiles and config.

`./container/legacy/` is dedicated to legacy architectures like armv7. Related Dockerfiles and configs go there.

inetol committed:
move func requirements at start

return42 commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+    local variant
+    local platform
+
+    # Check if git is installed
+    if ! command -v git &>/dev/null; then
+        die 1 "Git is not installed"
+    fi
```
No need for, but if it fits to your needs and you want: there is a `required_commands git` ..

https://github.com/searxng/searxng/blob/1b787ed35e9c51e335c42faee1f76695780ba4cb/utils/lib.sh#L60-L74

inetol replied:
a22df071107a67604f1d881cc145af0f27246f2c

return42 commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+    info_msg "Selected engine: $container_engine"
+
+    # Setup arch specific
+    case $parch in
+    "X64" | "x86_64" | "amd64")
+        dockerfile="Dockerfile"
```
nitpicking: indentation can be fixed to 8 spaces .. related to https://github.com/searxng/searxng/pull/4707#discussion_r2079256799 .. but we can also standardize indentations later when you create the PR for a shell formater.

return42 commented on utils/lib_sxng_container.sh:
```diff
@@ -0,0 +1,319 @@
+        if [ "$GITHUB_ACTIONS" = "true" ]; then
+            "$container_engine" push "ghcr.io/$CONTAINER_IMAGE_ORGANIZATION/cache:$CONTAINER_IMAGE_NAME-$arch$variant"
+
+            # Output to GHA
+            {
+                echo "version_string=$VERSION_STRING"
+                echo "version_tag=$VERSION_TAG"
+                echo "docker_tag=$DOCKER_TAG"
+                echo "git_url=$GIT_URL"
+                echo "git_branch=$GIT_BRANCH"
+            } >>"$GITHUB_OUTPUT"
```
No need for, just a hind .. for multiline outputs I often use *heredoc format*

```bash
cat << EOF > /path/to/your/file
This line will write to the file.
${THIS} will also write to the file, with the variable contents substituted.
EOF
```

*heredoc* is quite good to *read* and comfortable to *write*  .. it can be used in a variety of ways ..

If you want to pipe a series of commands e.g. to a bash .. and these commands should be displayed simultaneously on the terminal (to show the progress of the script)

https://github.com/searxng/searxng/blob/1b787ed35e9c51e335c42faee1f76695780ba4cb/utils/lib.sh#L893-L896

- https://linuxize.com/post/bash-heredoc/
- https://tldp.org/LDP/abs/html/here-docs.html

inetol replied:
b43428830cb845dd752a41a6bd910e60dc23253d
